### PR TITLE
Add Delaware combined separate filing status

### LINKED ISSUE #3349: Add Delaware combined separate filing status
In the 0.551.0 version, the Delaware income tax is incomplete in that it does not include the option for a married couple to file combined separate on the same state return.  The Delaware income tax form shows the Column A and Column B to permits this, and the instructions make it clear that it is advantageous for couples to file combined separate returns in many cases (see the "**Note: Generally ...**" text in the following screenshot from page 5 of the [2022 instructions](https://revenuefiles.delaware.gov/2022/PIT-RES_TY22_2022-02_Instructions.pdf#page=5)).

Screenshot 2023-12-01 at 4 13 09 PM

Failing to model DE Filing Status 4 causes Policyengine to overestimate DE income tax in thousands of cases when comparing with TAXSIM35, which does calculate to see if the combined separate filing status is advantageous for a married couple.

For ideas on how to do these combined separate calculations for Delaware, study the Iowa income tax code in Policyengine.

### PR DESCRIPTION
**Pull request by xianxinzhu**
Fixes #3349

### CODE REVIEW

File: policyengine_us/parameters/gov/states/hi/tax/income/rates/single.yaml [hunk 1]
```diff
@@ -25,11 +25,11 @@ brackets:
     rate:
       2017-01-01: 0.072
   - threshold:
-      2017-01-01: 24_000
+      2017-01-01: 36_000
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/hi/tax/income/rates/single.yaml:28]: this change is on purpose right?

File: policyengine_us/variables/gov/states/hi/tax/income/hi_income_tax_before_non_refundable_credits.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/hi/tax/income/hi_income_tax_before_non_refundable_credits.py:1]: add these files back

File: policyengine_us/variables/gov/states/tax/income/state_income_tax.py [hunk 2]
```diff
@@ -20,14 +20,14 @@ class state_income_tax(Variable):
...
         # "ga_income_tax",  --- activating will cause circular logic errors
-        # "hi_income_tax",  --- activating will cause circular logic errors
+        "hi_income_tax",
         # "ia_income_tax",  --- activating will cause circular logic errors
         "id_income_tax",
         "il_income_tax",
         "in_income_tax",
         "ks_income_tax",
         "ky_income_tax",
-        # "la_income_tax",  --- activating will cause circular logic errors
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/tax/income/state_income_tax.py:30]: revert changes

File: policyengine_us/variables/household/demographic/marital_unit/is_return_combined.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from policyengine_us.model_api import *
+
+
+class is_return_combined(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/household/demographic/marital_unit/is_return_combined.py:4]: delete this file, we have de_can_file_separate_on_same_return

File: policyengine_us/variables/gov/states/de/tax/income/additions/de_additions.py [hunk 4]
```diff
@@ -3,8 +3,8 @@
 
 class de_additions(Variable):
     value_type = float
-    entity = TaxUnit
-    label = "Delaware agi additions"
+    entity = Person
+    label = "Delaware individual additions"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/additions/de_additions.py:7]:
```suggestion
    label = "Delaware adjusted gross income additions"
```

File: policyengine_us/variables/gov/states/de/tax/income/de_agi.py [hunk 5]
```diff
@@ -3,11 +3,11 @@
 
 class de_agi(Variable):
     value_type = float
-    entity = TaxUnit
-    label = "Delaware adjusted gross income"
+    entity = Person
+    label = "Delaware individual adjusted gross income"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/de_agi.py:7]:
```suggestion
    label = "Delaware adjusted gross income"
```

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: we need tests for all new variables -> any formula needs a test

File: policyengine_us/parameters/gov/states/hi/tax/income/rates/single.yaml (hunk 1 above)
Comment (xianxinzhu) [reply policyengine_us/parameters/gov/states/hi/tax/income/rates/single.yaml:28]: I remember I added 24000 here... According to the merged version, it should be 24000.

File: changelog_entry.yaml
Comment (xianxinzhu) [reply changelog_entry.yaml:1]: Working on it.

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_eligible_indv.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+from policyengine_us.model_api import *
+
+
+class de_elderly_or_disabled_income_exclusion_eligible_indv(Variable):
+    value_type = bool
+    entity = Person
+    label = "Individual Eligibility for the Delaware elderly or disabled income exclusion"
+    definition_period = YEAR
+    defined_for = "de_can_file_separate_on_same_return"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_eligible_indv.py:9]:
```suggestion
    defined_for = StateCode.DE 
```

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_eligible_indv.py:9]: we want to move away from this definition its causing issues internally

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_eligible_indv.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+from policyengine_us.model_api import *
+
+
+class de_elderly_or_disabled_income_exclusion_eligible_indv(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_eligible_indv.py:4]:
```suggestion
class de_elderly_or_disabled_income_exclusion_eligible_person(Variable):
```

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_eligible_indv.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    value_type = bool
+    entity = Person
+    label = "Individual Eligibility for the Delaware elderly or disabled income exclusion"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_eligible_indv.py:7]:
```suggestion
    label = "Eligible person for the Delaware elderly or disabled income exclusion"
```

File: policyengine_us/variables/gov/states/de/tax/income/de_can_file_separate_on_same_return.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/de_can_file_separate_on_same_return.py:1]: delete file

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_capped_real_estate_tax.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+        "https://delcode.delaware.gov/title30/c011/sc02/index.html",
+    )
+    defined_for = StateCode.DE
+
+
+def formula(tax_unit, period, parameters):
+    p = parameters(period).gov.irs.deductions
+    filing_status = tax_unit("filing_status", period)
+    real_estate_tax = add(tax_unit, period, ["real_estate_taxes"])
+    capped_real_estate_tax = min_(
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_capped_real_estate_tax.py:21]:
```suggestion
    return min_(
```

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_eligible_indv.py (hunk 6 above)
Comment (xianxinzhu) [reply policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_eligible_indv.py:9]: Do you mean that we need to move away all that are defined for "de_can_file_seperate_on_same_return" and replace them with "StateCode.DE" or just this varibable?

Comment (xianxinzhu) [reply policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_eligible_indv.py:9]: But if we remove this file, how can we distinguish between the combined separate and joint filing status? Say, for the elderly_or_disabled_income_exclusion, it follows the logic in the following table. Now we calculate the exclusion for couples filing combined separately in `de_elderly_or_disabled_income_exclusion_eligible_person`, and do the calculation for the joint filing status couple in `de_elderly_or_disabled_income_exclusion_eligible`. But if we remove this defined for restriction, how could we realize such calculation respectively for combined separate and joint filing status? 

image

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_joint.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_joint.py:1]: is this variable used anywhere in the income tree (do we use this output for any other computation?

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_joint.py:1]: if not we can rename de_pension_exclusion_indv to `de_pension_exclusion` and keep it person level in the sum of subtractions

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_joint.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/subtractions/exclusions/de_elderly_or_disabled_income_exclusion_joint.py:1]: same question and comment as for the pension exclusion

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: could you add two integration tests - the way we have it in the Kentucky PR (the file called integration.yaml but with de outputs)

File: policyengine_us/parameters/gov/states/de/tax/income/deductions/itemized/sources.yaml [hunk 10]
```diff
@@ -0,0 +1,17 @@
+description: Delaware lets filers itemize the following deductions.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/de/tax/income/deductions/itemized/sources.yaml:1]:
```suggestion
description: Delaware accounts for the following itemized deductions.
```

File: policyengine_us/variables/gov/states/de/tax/income/de_pre_exclusions_agi.py [hunk 11]
```diff
@@ -3,11 +3,11 @@
...
-    entity = TaxUnit
-    label = "Delaware adjusted gross income before exclusions"
+    entity = Person
+    label = "Delaware individual adjusted gross income before exclusions"
     unit = USD
     definition_period = YEAR
     defined_for = StateCode.DE
 
-    adds = ["de_additions", "adjusted_gross_income"]
+    adds = ["de_additions", "adjusted_gross_income_person"]
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/de_pre_exclusions_agi.py:12]: could you do the same thing here as we do in KY and AR for the AGI file 
-> allocate the income of the dependents to the head of the household

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/de/tax/income/de_pre_exclusions_agi.py:12]: write a formula and create a unit test with a head + spouse + dependent with income

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_joint.py
Comment (xianxinzhu) [reply policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_joint.py:1]: I see that the current sum in the subtraction is in person level. Do we still need the de_pension_exclusion_joint?

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_indv.py [hunk 12]
```diff
@@ -1,10 +1,10 @@
 from policyengine_us.model_api import *
 
 
-class de_pension_exclusion(Variable):
+class de_pension_exclusion_indv(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_indv.py:4]:
```suggestion
class de_pension_exclusion(Variable):
```

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_joint.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_joint.py:1]: delete this file

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_indv.py [hunk 13]
```diff
@@ -1,10 +1,10 @@
...
     value_type = float
-    entity = TaxUnit
-    label = "Delaware pension exclusion"
+    entity = Person
+    label = "Delaware individual pension exclusion when married couples filing separately"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion_indv.py:7]:
```suggestion
    label = "Delaware individual pension exclusion"
```

File: policyengine_us/parameters/gov/states/de/tax/income/subtractions/subtractions.yaml [hunk 14]
```diff
@@ -1,7 +1,7 @@
 description: Delaware subtracts these items when calculating state adjusted gross income.
 values:
   2021-01-01:
-    - de_pension_exclusion # (4)
+    - de_pension_exclusion_indv # (4)
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/de/tax/income/subtractions/subtractions.yaml:4]:
```suggestion
    - de_pension_exclusion # (4)
```

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_capped_real_estate_tax.py [hunk 15]
```diff
@@ -16,17 +16,8 @@ class de_itemized_deductions(Variable):
...
-            deduction
-            for deduction in p.itemized_deductions
-            if deduction not in ["salt_deduction"]
-        ]
-        federal_deductions = add(tax_unit, period, deductions)
-
         real_estate_tax = add(tax_unit, period, ["real_estate_taxes"])
-
-        capped_real_estate_tax = min_(
+        print(real_estate_tax)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_capped_real_estate_tax.py:20]: Remove

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_capped_real_estate_tax.py [hunk 16]
```diff
@@ -16,17 +16,8 @@ class de_itemized_deductions(Variable):
...
+        return min_(
             real_estate_tax, p.itemized.salt_and_real_estate.cap[filing_status]
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_capped_real_estate_tax.py:22]: Separate cap as its own python variable (one operation per line)

File: policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion.py [hunk 17]
```diff
@@ -71,4 +69,4 @@ def formula(tax_unit, period, parameters):
             )
         else:
             exclusion_amount = capped_eligible_pension_income
-        return tax_unit.sum(exclusion_amount)
+        return exclusion_amount
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/de/tax/income/subtractions/pension_exclusion/de_pension_exclusion.py:72]: Move the return within if-else to save the extra line

File: policyengine_us/variables/gov/states/de/tax/income/de_income_tax_before_non_refundable_credits_indv.py [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+    value_type = float
+    entity = Person
+    label = "Delaware personal income tax before non-refundable credits when married filing separately"
+    unit = USD
+    definition_period = YEAR
+    defined_for = StateCode.DE
+
+    def formula(person, period, parameters):
+        taxable_income = person("de_taxable_income_indv", period)
+        return parameters(period).gov.states.de.tax.income.rate.calc(
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/de/tax/income/de_income_tax_before_non_refundable_credits_indv.py:14]: Make p the parameter folder to reduce wrapping and align with convention

**Review by MaxGhenis — APPROVED** [review]
Good to merge after these nits

**Pull request merged**